sympy/integrals/integrals.py

```diff
@@ -1402,7 +1410,19 @@ def principal_value(self, **kwargs):
             I += limit(((F.subs(x, s - r)) - F.subs(x, s + r)), r, 0, '+')
         return I
 
-
+def recur_free_in(obj, syms, bs):
+    if obj in syms and obj not in bs:
+        yield obj
+    else:
+        if isinstance(obj, Tuple):
+            if len(obj) == 1:
+                yield from recur_free_in(obj[0], syms, bs)
+        for arg in reversed(obj.args[1:]):
+            if obj.args[0] not in bs:
```

Comment:
It is best to use `obj.limits` and `obj.function` rather than accessing `obj.args` directly.

Reply:
The issue with using `obj.limits` is that as we recursively call the limits, thus `obj` will not remain an `Integral` object after first calling. 
For e.g. in `Integral(x, (x, 0, x)).has_free({x})`, obj will be 
```
Integral(x, (x, 0, x))
(x, 0, x)
x
```
Here in `(x, 0, x)` `obj` is no longer an `Integral` object.